### PR TITLE
[JENKINS-59844] Process Launcher shouldn't throw ArrayIndexOutOfBoundException on empty command

### PR DESCRIPTION
**Pull request by EstherAF**
See [JENKINS-59844](https://issues.jenkins-ci.org/browse/JENKINS-59844).

### Proposed changelog entries

* Entry 1: internal: Log when Launcher is invoked with an empty command, informing that it can throw an unexpected exception

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developers, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention @MRamonLeon

### CODE REVIEW

File: core/src/main/java/hudson/Launcher.java [hunk 1]
```diff
@@ -450,8 +450,14 @@ public ProcStarter writeStdin() {
 
         /**
          * Starts the new process as configured.
+         *
+         * @throws RuntimeException can be thrown if {@link #cmds()} is empty
          */
         public Proc start() throws IOException {
+            if(cmds().isEmpty()){
```
Comment (daniel-beck) [inline core/src/main/java/hudson/Launcher.java]:
```suggestion
            if (cmds().isEmpty()) {
```

File: core/src/main/java/hudson/Launcher.java [hunk 2]
```diff
@@ -450,8 +450,14 @@ public ProcStarter writeStdin() {
 
         /**
          * Starts the new process as configured.
+         *
+         * @throws RuntimeException can be thrown if {@link #cmds()} is empty
```
Comment (daniel-beck) [inline core/src/main/java/hudson/Launcher.java]: "Can" be thrown? When would it not be thrown?

File: core/src/main/java/hudson/Launcher.java [hunk 3]
```diff
@@ -450,8 +450,14 @@ public ProcStarter writeStdin() {
...
          */
         public Proc start() throws IOException {
+            if(cmds().isEmpty()){
+                LOGGER.log(Level.INFO, "Trying to start a process without setting a command. This might cause "
+                                       + "an unexpected exception");
```
Comment (daniel-beck) [inline core/src/main/java/hudson/Launcher.java]: Seems to be we want the stack trace here, otherwise how does this even help?

File: test/src/test/java/hudson/ProcStarterTest.java [hunk 4]
```diff
@@ -45,6 +45,7 @@
 import hudson.tasks.BuildWrapperDescriptor;
 import hudson.tasks.Builder;
 
+
```
Comment (daniel-beck) [inline test/src/test/java/hudson/ProcStarterTest.java]: Unnecessary whitespace change is no big deal here, but it's generally better to minimize occurrences of something like this before you submit a PR.

File: core/src/main/java/hudson/Launcher.java (hunk 3 above)
Comment (EstherAF) [reply core/src/main/java/hudson/Launcher.java]: The exception is not being caught.
Logging or not the exception is up to the client. If they want to handle it, logging it will only add noise to the logs. If they don't handle it, it's a RuntimeException, at some point, it will appear at the logs.

I'm detailing the exception a little bit thought

File: core/src/main/java/hudson/Launcher.java (hunk 2 above)
Comment (EstherAF) [reply core/src/main/java/hudson/Launcher.java]: Depending on the Launcher implementation. DummyLauncher will never throw it. Is it worth clarifying this? or can we just generalize and say that "will be thrown if cmds() is empty"?

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
Thanks for the pull request! Unfortunately I do not undertand what is the end goal. The ticket and the PR title say "Process Launcher shouldn't throw ArrayIndexOutOfBoundException on empty command" while here we merely have extra logging here.

Could we just fail commands if it is not permitted?

**Comment by daniel-beck** [comment]
It is unclear what the benefit of additional logging is, especially given so little information that the admin cannot do anything about it anyway.

From recent issue comments:

>  This should help any user to track down the cause of this exception in the future.

A Jenkins instance often has hundreds, or thousands, of jobs. How would they even identify which build would be about to fail? This isn't printed in the job log (which will at least have that context), but the global Jenkins log.

If it's not a build in the first place, it's probably even worse…

**Comment by EstherAF** [comment]
> Could we just fail commands if it is not permitted?

That was my first thought, and my prefered solution, but after running this by @MRamonLeon I think that this change can break other plugins compatibility if they are already handling this exception, despite of not being documented. 

If this is not a problem, I'm happy to implement this approach.

> A Jenkins instance often has hundreds, or thousands, of jobs. How would they even identify which build would be about to fail? This isn't printed in the job log (which will at least have that context), but the global Jenkins log.
> 
> If it's not a build in the first place, it's probably even worse…

We didn't detect this in a job, but in a plugin that was launching a command (customizable by the user) to check if the node was eligible to run jobs. 
Currently, there is no information at all about this behaviour, but with the javadoc the developers know about it, and with the log, if the exception is not being handled, at least they have a  little clue about why is happening.

Though, I agree with @oleg-nenashev that the best approach would be just handling this behaviour as a failed command. If you agree, I'll change the PR.

**Comment by oleg-nenashev** [comment]
> That was my first thought, and my prefered solution, but after running this by @MRamonLeon I think that this change can break other plugins compatibility if they are already handling this exception, despite of not being documented.

Do we have examples of plugins which are about to break?

**Comment by MRamonLeon** [comment]
The goal of the PR was to help to diagnose why this happens without changing the interface (exceptions thrown included) of such a used class (`hudson.Launcher`). 

Another approach was to add the check in the `LocalProc` constructor before calling the `procBuilder.start()` which will fail for sure with an empty string, but we agreed to be conservative avoiding returning a different exception and just help with a message in the log.

**Review by MRamonLeon — APPROVED** [review]

File: core/src/main/java/hudson/Launcher.java [hunk 5]
```diff
@@ -450,8 +450,14 @@ public ProcStarter writeStdin() {
 
         /**
          * Starts the new process as configured.
+         *
+         * @throws ArrayIndexOutOfBoundsException when {@link #cmds()} is empty
          */
         public Proc start() throws IOException {
+            if (cmds().isEmpty()) {
+                LOGGER.log(Level.INFO, "Trying to start a process without setting a command. This might cause "
```
Comment (rsandell) [inline core/src/main/java/hudson/Launcher.java]: Should be `Level.WARNING` I think

**Comment by varyvol** [comment]
@daniel-beck @EstherAF if the stack trace for the ArrayOutOfBounds exception is logged, would that provide useful information to the admin? If so, the exception could be captured, logged and re-thrown.

File: core/src/main/java/hudson/Launcher.java [hunk 6]
```diff
@@ -450,8 +450,14 @@ public ProcStarter writeStdin() {
...
         /**
          * Starts the new process as configured.
+         *
+         * @throws ArrayIndexOutOfBoundsException when {@link #cmds()} is empty
          */
         public Proc start() throws IOException {
+            if (cmds().isEmpty()) {
+                LOGGER.log(Level.WARNING, "Trying to start a process without setting a command. This might cause "
+                                       + "an ArrayIndexOutOfBounds exception");
+            }
```
Comment (fcojfernandez) [inline core/src/main/java/hudson/Launcher.java]: Is there a way to know what is going to launch the empty process. It would be great to have that information. If not, this message would not help to diagnose what is/will happening

Comment (MRamonLeon) [reply core/src/main/java/hudson/Launcher.java]: There is no command to launch, "".

**Comment by MRamonLeon** [comment]
@varyvol The problem is that an `ArrayOutOfBounds` is not very intuitive. That's the reason behind this PR. It took us a long time to find out the problem in a recent report from one of our instances.

**Review by rsandell — APPROVED** [review]

**Comment by oleg-nenashev** [comment]
I think it needs more discussion before merging @MRamonLeon @rsandell

**Comment by varyvol** [comment]
@EstherAF to be honest, after the decision of not breaking compatibility (which I understand) I don't think this provides that much value. That seems to be the general feeling as well, so I'm marking this as "proposed for close". Please feel free to let us know in case you don't agree.
